### DISCUSSION TITLE
Wayland cursor capture

### DISCUSSION THREAD

**Discussion by rustdesk** [rustdesk/rustdesk#5403] [Team]
https://github.com/rustdesk/rustdesk/blob/309cb1010ce9762cfd334da8d87b9b3328194b58/src/platform/linux.rs#L113 this works on X11 only, wayland implementation not done yet.

**Comment by rustdesk** [comment]
Any idea / proposal about this? @sahilyeole

**Comment by beelchester** [comment]
Has this crate been tried before?
[wayland_cursor](https://docs.rs/wayland-cursor/latest/wayland_cursor/) (along with [wayland_client](https://docs.rs/wayland-client/0.30.0/wayland_client/index.html) for `Connection` and `WlShm`)

If not, I think it might help us. However, I have not tested it yet (I'm still figuring out how to get `WlShm` for it).
But what do you think, should I go ahead and try to implement it?

**Reply by rustdesk** [reply]
I checked before, It more likely load a cursor image and show it, rather than capture current cursor.

**Reply by beelchester** [reply]
Ah, ok. I need to dig more.

**Reply by rustdesk** [reply]
Check if chrome / qt etc has some implementation.

**Reply by beelchester** [reply]
Here's my understanding till now (I may be wrong though):
- Wayland seems to be very restricting than x11. I could not find any global access to the cursor/pointer data for wayland yet.
- Wayland protocol provides pointer data and methods only if it's within the surface specified by the application, not outside of it. I saw multiple implementations on github using a similar concept.
- However, for our use case, we require global cursor data regardless of the pointer's position.

I'm still working on it, let's see if I can find anything meaningful soon.

**Reply by beelchester** [reply]
https://github.com/libsdl-org/SDL/pull/7693
I'm not sure if something like this will help us or not.

> Wayland doesn't support getting the true global cursor position, but it can be faked

> this gives 'correct' coordinates **_when a window has focus_**, which is good enough for most applications.

**Comment by rustdesk** [comment]
Progress?

**Reply by rustdesk** [reply]
https://community.teamviewer.com/English/discussion/122410/teamviewer-support-on-wayland-experimental-state
![image](https://github.com/rustdesk/rustdesk/assets/71636191/4c3ea82a-202b-475f-8404-e152326af1a9)

**Reply by rustdesk** [reply]
Wayland protocol does not support these. We have to do on top of composor. Let's try out the default wayland composor of Ubuntu 22.04 first. If we can not get cursor image, we can also try out to get cursor name (then draw ourself).

**Reply by beelchester** [reply]
Ok, Thanks for helping. I will do that. 👍

**Reply by rustdesk** [reply]
Another "stupid" way (not sure if it is feasible, or may be inefficient) is, we know we can get the cursor position, we know we can capture part of the screen with cursor and without cursor, so we can extract the cursor from the difference of the two picture.

**Reply by rustdesk** [reply]
Be clever! Young man, :).

**Reply by beelchester** [reply]
Ah, interesting. Can try thanks!

**Reply by rustdesk** [reply]
Let's try composor first.

**Reply by beelchester** [reply]
Ok

**Comment by beelchester** [comment]
Pipewire's metadata seems to be the answer.
https://docs.pipewire.org/structspa__meta__cursor.html

obs has its implementation. Figuring out how we can implement it as well.

**Comment by rustdesk** [comment]
@fufesou what is the problem of obs implementation? we tried out before, and then paused.

**Reply by fufesou** [reply]
obs can grab the cursor.
But the implementation is a little complicated.
It implements the cursor transfer of the stream pipewire. And it depends on some low lever OPENGL APIs.

**Reply by beelchester** [reply]
How does `capture_cursor` & `cursor_mode` implementation work here?https://github.com/rustdesk/rustdesk/blob/f305c9d96afa1c55b60a3dba521f85745b009ce8/libs/scrap/src/wayland/pipewire.rs#L607

**Reply by rustdesk** [reply]
cursor_mode has nothing to do with standalone cursor capture.

**Reply by rustdesk** [reply]
I mentioned cursor_mode last time.

> we know we can capture part of the screen with cursor and without cursor, so we can extract the cursor from the difference of the two picture.

**Comment by beelchester** [comment]
Hi @fufesou,
If you recall, could you please give some insights on this and why it was disabled later (cursor embed)?
https://github.com/rustdesk/rustdesk/pull/2360
https://github.com/rustdesk/rustdesk/pull/3718

Also, did you previously attempt to set the cursor mode value to 4 and extract cursor metadata from the stream?
(According to https://flatpak.github.io/xdg-desktop-portal/#gdbus-property-org-freedesktop-portal-ScreenCast.AvailableCursorModes)
As you have commented this:
https://github.com/rustdesk/rustdesk/blob/d688e345216247928c2f7ff0f8eea49072e6cbc2/libs/scrap/src/wayland/pipewire.rs#L495

**Reply by fufesou** [reply]
The embedded cursor will cause a large delay in the operation experience.

It is able to set the cursor mode value to 4. And the obs use this cursor mode.
But the gstreamer crate ( https://crates.io/crates/gstreamer )  does not support  the cursor data transfer.
You have to implement the cursor data transfer by yourself if you want to use the cursor mode 4.

**Reply by beelchester** [reply]
Oh, ok. Thanks.

**Reply by beelchester** [reply]
@fufesou Sorry for interrupting you again, but do I need to completely rewrite our wayland screen capture into low-level pipewire code, something similar to obs?

With our current implementation, I could not find any way to directly access the pipewire stream or buffer, which would allow us to utilize pipewire api methods on it.
Also, I'm not sure if it's possible to specifically transfer only the cursor metadata, it might require complete rewriting.

Unfortunately, obs approach is the only good solution I could find for capturing the cursor :(

**Reply by fufesou** [reply]
Ok. You may want to refer to these source code

https://github.com/PipeWire/pipewire
https://github.com/GStreamer/gstreamer
https://gitlab.freedesktop.org/gstreamer/gstreamer-rs

**Reply by beelchester** [reply]
Ok, Thank you so much.

**Comment by beelchester** [comment]
@fufesou  Are you aware of or have any clue about these wayland behaviours?

1. Sometimes the cursor image is shown on ubuntu wayland instead of a blue circle. It doesn't seem to be embedded because it's roughly only 70% of the time correct as the actual cursor image on wayland side.

<img width="281" alt="Screenshot 2023-09-16 at 6 33 48 PM" src="https://github.com/rustdesk/rustdesk/assets/73148455/4ffa44ab-4b13-433d-a9e8-29a4b954a584">

2. Input works on the official release but on `cargo run` it doesn't work (keyboard, mouse both). It worked once or twice but not after that. The official release which was earlier working also stops taking input. Reinstalling rustdesk didn't work had to reinstall ubuntu and then again after `cargo run` same behaviour occurred :(

Also, when 2 happens the cursor image is always shown.

**Reply by fufesou** [reply]
For 1, it's a long time that I haven't changed the wayland cursor. I briefly looked at the code, but I couldn't find where to handle the wayland cursor.

For 2, you need to run `--service` as the root process and `--server` as the user process to make the input working.
Uinput can only be handled in the root priviledge. The wayland screen capturing should be in the user process.

**Reply by beelchester** [reply]
Alright, working. Thanks for the help :)

I'll look something up for 1 later. Maybe just improving its implementation can make wayland cursor capture work well and consistently.

**Comment by FurtadoPires** [comment]
Sorry I'm not a dev, but I think this might be interesting for you guys

https://github.com/flatpak/xdg-desktop-portal/releases/tag/1.18.0

> Highlights
> 
>     A new config-based portal matching mechanism that gives more precise control over which portal backends are picked for each portal.
>     New portals: Clipboard and Input Capture
>     The settings portal now documents an 'accent-color' key

**Reply by beelchester** [reply]
Hey, Thanks for sharing.
But it's about input capturing, we already have that in wayland.
We are interested in capturing the cursor metadata i.e. size and image of the cursor.

**Comment by beelchester** [comment]
>Sometimes the cursor image is shown on ubuntu wayland instead of a blue circle.
>Maybe just improving its implementation can make wayland cursor capture work well and consistently.

The cursor capture is already working because `XFixesGetCursorImage` is available in wayland due to Xwayland. Checked on ubuntu wayland.
https://github.com/rustdesk/rustdesk/blob/230eb7653211d5f6199c9dbe3f9896be2fe9597c/src/platform/linux.rs#L119

But, the cursor capture fails sometimes (showing blue circle instead). So, I'm debugging that.

**Reply by rustdesk** [reply]
Update?

**Reply by beelchester** [reply]
The cursor capture failed for me only when `--service` was started on wayland login screen.
But the cursor capture was working if I restarted `--service` manually after logging in and then making a remote connection.

Let me know if I'm wrong, but here's my view:
On walyand login screen when `--service` starts it tries to make a connection/pointer to xdisplay but it fails hence `get_cursor()` fails.
https://github.com/rustdesk/rustdesk/blob/3e61b894996b714665af0925957f29116b7ff9e8/src/platform/linux.rs#L36
But if I restart `--service` after logging in the connection to xdisplay can be made successfully and hence `get_cursor()` works.

Potential solutions I'm trying: (only for wayland)
1. restart `--service` after user has logged in.
2. reinitialize thread local `DISPLAY` variable after user has logged in.

**Reply by fufesou** [reply]
@sahilyeole I've tested wayland on Ubuntu 22.04. Restarting `--service` still cannot make the cursor capturer work.

But if I open a rustdesk window immediately after running `systemctl restart rustdesk`, and then connect, the cursor capturer will work.
`ps -ef | grep rustdesk` shows `--server` is not running at this time.
There's a inner server thread in the rustdesk ui at this time.
![1695905043893](https://github.com/rustdesk/rustdesk/assets/13586388/ac8b3071-0b4b-4609-b035-060700110f1a)



If I wait the `--server` started by `--service`, the cursor cannot be captured.

![1695904947515](https://github.com/rustdesk/rustdesk/assets/13586388/96e22287-4e78-4790-b935-217557be2cff)

**Reply by fufesou** [reply]
I'm not sure if it is the same to you.
Maybe you need to solve the permission or ENV problem.


> reinitialize thread local DISPLAY variable after user has logged in.

It seems you're already trying to do this.

**Reply by beelchester** [reply]
Thanks, @fufesou.
That's exactly what I needed.
Earlier, I was trying the build before https://github.com/rustdesk/rustdesk/pull/5809. The `--server` was starting late that's why if I made the connection before `--server` the cursor was captured.
But now I tried the latest build, cursor capture is not working because `--server` is starting very quickly.
Thanks, for pointing that out :)

**Reply by rustdesk** [reply]
update?

**Reply by beelchester** [reply]
I was working on https://github.com/rustdesk/rustdesk/issues/5853, I will get back on this soon.

**Reply by beelchester** [reply]
>inner server thread in the rustdesk ui

@fufesou can you please link me to the part of the code responsible for inner server thread?
Sorry to disturb but I'm not able to find it.

---
Till now I tried reinitializing DISPLAY variable, and restarting --server when DISPLAY is null, i.e. connection to XOpenDisplay is null
 https://github.com/rustdesk/rustdesk/blob/3e61b894996b714665af0925957f29116b7ff9e8/src/platform/linux.rs#L36
 But it is giving null all the time. It is only working in the inner server thread scenario, when --server has yet to start.

**Reply by fufesou** [reply]
Server thread

The inner thread is started in this function

https://github.com/rustdesk/rustdesk/blob/3e61b894996b714665af0925957f29116b7ff9e8/src/core_main.rs#L164

---------------------------------------------------------

Server process

For wayland, the "--server" subprocess is started here

https://github.com/rustdesk/rustdesk/blob/3e61b894996b714665af0925957f29116b7ff9e8/src/platform/linux.rs#L387

For X11, the "--server" subprocess is started here

https://github.com/rustdesk/rustdesk/blob/3e61b894996b714665af0925957f29116b7ff9e8/src/platform/linux.rs#L370

**Comment by fufesou** [comment]
The wayland clipboard on the controlled side also do not work if `--server` is started inappropriately.

https://github.com/rustdesk/rustdesk/pull/5900

**Reply by beelchester** [reply]
Does wayland clipboard require root permission?

**Reply by fufesou** [reply]
No, the `--server` is always the user permission on Wayland.

**Comment by rustdesk** [comment]
> Update:
> working on the wayland cursor, today I spent many hours debugging the issue. The connection/pointer to xopendisplay is important in order to draw cursor.
> I tried making this connection in multiple scenarios such as root --server, start with --service, etc but didn't work.. but one thing I'm clear now is that it's only not null in the inner UI server code and then it is overridden by --server in which connection is null irrespective of root or user.
> It's taking time but I'm giving it my best and I think I'm getting closer to the solution.

Are you sure you are doing in the correct way? Since you mentioned it worked before https://github.com/rustdesk/rustdesk/pull/5809.

**Reply by beelchester** [reply]
Hopefully, I'm doing it correctly. It worked before #5809 because `--server` was starting late and I was making attempts before `--server` started, i.e. on the inner UI server. Even before #5809 if I wait and let `--server` start then the cursor capture fails.
Now, `--server` starts immediately and hence cursor capture is failing almost all the time.

**Reply by rustdesk** [reply]
> Even before https://github.com/rustdesk/rustdesk/pull/5809 if I wait and let --server start then the cursor capture fails.

Any clue about this?

**Reply by beelchester** [reply]
I think mainly `--server` is causing it.
So, this:
>connection/pointer to xopendisplay
>it's only not null in the inner UI server code and then it is overridden by --server in which connection is null irrespective of root or user.

---
I tried making some changes to `--server` which is causing it to crash and restart repeatedly and the cursor capture is working, when `--server` isn't working. So, most likely improving the `--server` configuration for wayland will fix it (trying to figure out this part).

https://github.com/rustdesk/rustdesk/assets/73148455/4e660486-47cf-4563-a6f6-bd672aa17382

**Reply by rustdesk** [reply]
Let's see the code.

**Comment by beelchester** [comment]
I'm not sure if this solution is acceptable, but I can compile and run a c code separately which will handle the XOpenDisplay connection and XFixesGetCursorImage and then give its output to `get_cursor()` and `get_cursor_data()` for cursor capturing.

Below is the working c code for printing the latest cursor serial on loop (tested on wayland), I can use its output for `get_cursor()` and I can also try to modify it for `get_cursor_data()`:
```c
#include <stdio.h>
#include <stdlib.h>
#include <X11/Xlib.h>
#include <X11/extensions/Xfixes.h>
#include <unistd.h>

int main() {
    Display *display = XOpenDisplay(NULL);
    if (!display) {
        fprintf(stderr, "Failed to open X display\n");
        return 1;
    }
    long previous_serial = -1;
    while (1) {
        XFixesCursorImage *img = XFixesGetCursorImage(display);
        if (img) {
            if (img->cursor_serial != previous_serial) {
                printf("Cursor serial: %ld\n", img->cursor_serial);
                previous_serial = img->cursor_serial;
            }
            XFree(img);
        } else {
            fprintf(stderr, "Failed to get cursor image\n");
        }
        sleep(0.5); 
    }
    XCloseDisplay(display);
    return 0;
}
```

Let me know if it's acceptable or not.

**Reply by fufesou** [reply]
I think you need to find out how to capture the cursor on Wayland, run as user, not run it directly.

**Comment by rustdesk** [comment]
@fufesou take over this issue, @sahilyeole is not capable of this issue. This is high priority since I plan to release this in 1.2.3.

**Comment by rustdesk** [comment]
https://andri.yngvason.is/making-a-wayland-screen-capturing-protocol.html wayland will support cursor capture soon.
